### PR TITLE
ReplicatedPG: Don't cache recovery data

### PR DESCRIPTION
**Pull request by yuyuyu101**
Signed-off-by: Haomai Wang <haomaiwang@gmail.com>

### CODE REVIEW

**Comment by yuyuyu101** [comment]
If backfilling happened, recovery process will occur much of cache and influent online io. 

I don't sure whether it may influent other process?

**Comment by XinzeChi** [comment]
@yuyuyu101, I want to know if we cache this fd, and ceph recover it, the new patch will clear it?

**Comment by yuyuyu101** [comment]
Yeah, I think it will. This point seemed worth to consider.

**Comment by XinzeChi** [comment]
@yuyuyu101 , I think we can do more. Such as, we do not cache various recover object, for example objectcontext cache, fd cache and so on. What do you think?

**Comment by liewegas** [comment]
The write hint seems safe .. the OSD wasn't storing it before so we're clearly no worse off than before.

On the read side, I'm less sure. If we're a replica and the primary is pulling from us, dontneed makes sense.  But on the primary maybe not. Can we look at whether the obc was in the cache as a proxy for whether the object may have been warm or not?

**Comment by liewegas** [comment]
FWIW on the write/push side, we could pass a hint through perhaps that is set if there is a request that is blocking on this object's recovery, perhaps?

**Comment by XinzeChi** [comment]
@yuyuyu101 , I think we can do the same thing in be_deep_scrub. 

Maybe we can grep all the ceph code to do cleanup.

**Comment by liewegas** [comment]
lgtm

**Comment by athanatos** [comment]
Some comments, re-add needs-qa once addressed or disagreed with.

**Comment by yuyuyu101** [comment]
@athanatos I rethink the way to decide whether we invalidates page cache for recovery(read data).

I agree with your option and actually slave pg almostly doesn't need to cache data, so I think a simple here is invalidate page cache always in commit(https://github.com/yuyuyu101/ceph/commit/e6ff1831ba102826af1c4e676268b4b305aae05e). It should be simple and worked better.

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -397,6 +397,7 @@ void ReplicatedPG::wait_for_unreadable_object(
     if (is_missing_object(soid)) {
       recover_missing(soid, v, cct->_conf->osd_client_op_priority, h);
     } else {
+      h->cache = true;
```
Comment (athanatos) [inline src/osd/ReplicatedPG.cc]: cache = true here means that we should avoid caching the data?  Why?  We are in wait_for_unreadable_object, so there must be an IO pending.

File: src/osd/ReplicatedBackend.cc [hunk 2]
```diff
@@ -2078,8 +2086,8 @@ int ReplicatedBackend::build_push_op(const ObjectRecoveryInfo &recovery_info,
        p != out_op->data_included.end();
        ++p) {
     bufferlist bit;
-    store->read(coll, recovery_info.soid,
-		     p.get_start(), p.get_len(), bit);
+    store->read(coll, recovery_info.soid, p.get_start(), p.get_len(), bit,
+                cache ? CEPH_OSD_OP_FLAG_FADVISE_DONTNEED: 0);
```
Comment (athanatos) [inline src/osd/ReplicatedBackend.cc]: To me, it should be cache ? 0 : CEPH_OSD_OP_FLAG_FADVISE_DONTNEED.  That is, cache = true should mean that we are going to cache the data read.  I'd actually prefer that it had a more descriptive name like ignore_cache or cache_dont_need.

File: src/osd/ReplicatedBackend.cc [hunk 3]
```diff
@@ -1713,13 +1717,16 @@ void ReplicatedBackend::submit_push_data(
     t->omap_setheader(target_coll, recovery_info.soid, omap_header);
   }
   uint64_t off = 0;
+  uint32_t fadvise_flags = CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL;
+  if (!get_parent()->object_has_io_pending_recovery(recovery_info.soid))
```
Comment (athanatos) [inline src/osd/ReplicatedBackend.cc]: This isn't valid on the replica.  waiting_for_unreadable and waiting_for_degraded are only maintained on the primary.  You need to restructure how this information is passed around.  I'd also very much prefer that the call into the backend indicate whether the recovery IO should be marked as dont_need or not.  I don't really like poking back up into the parent to check the waiting_for_* queues.  The parent has the information available when it calls into the backend.

**Comment by athanatos** [comment]
Sorry, I got confused since you haven't cleaned up the previous commits yet.  I think what you are doing with the object_has_io_pending_recovery is better than using the object context cache, but still seems to be unnecessary.  For one thing, the replica does not maintain those sets.  For another, those callbacks into PG complicate the interface a lot.  I'd much rather the parent pass that information in when it initiates the recovery operations.

**Comment by yuyuyu101** [comment]
@athanatos I have cleaned up the commit, it should be much clean! Now I use PullIInfo to indicate we need to cache data.

**Comment by yuyuyu101** [comment]
ping @athanatos

File: src/osd/ECBackend.cc [hunk 4]
```diff
@@ -1778,6 +1778,10 @@ void ECBackend::be_deep_scrub(
   if (stride % sinfo.get_chunk_size())
     stride += sinfo.get_chunk_size() - (stride % sinfo.get_chunk_size());
   uint64_t pos = 0;
+
+  // Look at whether the obc was in the cache for whether the object is warm or not
```
Comment (liewegas) [inline src/osd/ECBackend.cc]: this comment is obsolete i think?

File: src/osd/ReplicatedBackend.cc [hunk 5]
```diff
@@ -742,13 +742,17 @@ void ReplicatedBackend::be_deep_scrub(
   bufferlist bl, hdrbl;
   int r;
   __u64 pos = 0;
+
+  // Look at whether the obc was in the cache for whether the object is warm or not
```
Comment (liewegas) [inline src/osd/ReplicatedBackend.cc]: here too

**Comment by liewegas** [comment]
passed my wip-sage-testing run, sage-2015-08-19_19:13:53-rados-wip-sage-testing-distro-basic-multi

**Pull request merged**